src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -53,66 +83,228 @@ static ucs_status_t uct_rdmacm_cm_query(uct_cm_h cm, uct_cm_attr_t *cm_attr)
+    remote_data.dev_addr              = dev_addr;
+    remote_data.dev_addr_length       = addr_length;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    listener->conn_request_cb(&listener->super, listener->user_data,
+                              dev_name, event, &remote_data);
+    ucs_free(dev_addr);
+}
 
-    if (rdma_connect(cep->id, &conn_param)) {
-        ucs_error("rdma_connect(to addr=%s) failed: %m",
+static void uct_rdmacm_cm_handle_event_connect_response(struct rdma_cm_event *event)
+{
+    struct sockaddr            *remote_addr = rdma_get_peer_addr(event->id);
+    uct_rdmacm_priv_data_hdr_t *hdr         = (uct_rdmacm_priv_data_hdr_t *)
+                                              event->param.conn.private_data;
+    uct_rdmacm_cm_ep_t         *cep         = event->id->context;
+    char                       ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    uct_device_addr_t          *dev_addr;
+    size_t                     addr_length;
+    uct_cm_remote_data_t       remote_data;
+    ucs_status_t               status;
+
+    ucs_assert(event->id == cep->id);
+
+    remote_data.field_mask            = UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    status = uct_rdmacm_cm_id_to_dev_addr(event->id, &dev_addr, &addr_length);
+    if (status != UCS_OK) {
+        ucs_error("client (ep=%p id=%p) failed to process a connect response "
+                  "from server %s.", cep, event->id,
                   ucs_sockaddr_str(remote_addr, ip_port_str,
                                    UCS_SOCKADDR_STRING_LEN));
+        uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, status);
+        return;
+    }
+
+    remote_data.field_mask       |= UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR |
+                                    UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR_LENGTH;
+    remote_data.dev_addr          = dev_addr;
+    remote_data.dev_addr_length   = addr_length;
+
+    uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, hdr->status);
+
+    ucs_free(dev_addr);
 
+    if (rdma_establish(event->id)) {
+        ucs_error("rdma_establish on ep %p (to server addr=%s) failed: %m",
+                  cep, ucs_sockaddr_str(remote_addr, ip_port_str,
+                                        UCS_SOCKADDR_STRING_LEN));
+        cep->disconnect_cb(&cep->super.super, cep->user_data);
     }
 }
 
+static void uct_rdmacm_cm_handle_event_rejected(struct rdma_cm_event *event)
+{
+    uct_rdmacm_cm_ep_t         *cep = event->id->context;
+    uct_rdmacm_priv_data_hdr_t *hdr;
+    uct_cm_remote_data_t       remote_data;
+
+    ucs_assert(event->id == cep->id);
+
+    /* Network reject or called by rdma_reject on the server */
+    remote_data.field_mask = 0;
+    if (event->param.conn.private_data != NULL) {
```

Comment:
assert that event->param.conn.private_data!= NULL (because server should always pass it)

Reply:
what if there was a reject from the network? not from an explicit API call from the server side

Reply:
yes, so still need to check